enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,445 @@
+the `CredentialRequest` CR. This resource allows other components access to the
+cloud API by creating a secret in the `kube-system` namespace which the resource
+issuing the `CredentialRequest` can fetch, containing the permissions required
+to communicate with the cloud API by the requesting component. The utilization
+of the CCO is dependent upon the questions in the [Summary](#summary), as it
+might be awkward deploying the CCO on a non-OpenShift cluster. This would be the
+case if an upstream development model is chosen.
+
+[10]: https://github.com/openshift/cloud-credential-operator
+
+##### RBAC
+
+Given the approach chosen above a separate set of RBAC permissions will be
+required for the cloud component / network plugins. These are listed below.
+
+Independently of the model chosen, the cloud component will require:
+
+WATCH/LIST/GET:
+
+- `egressips.k8s.ovn.org` - to watch for status assignments done by
+  OVN-Kubernetes
+- `netnamespace`/`hostsubnet` - to watch for the egress IP assignments request
+  for openshift-sdn
+- `secrets` - to watch for the creation of cloud credentials by the CCO
+- `nodes` - presumably to be able to verify node names / other info when
+  communicating with the cloud API
+
+**CRD**
+
+Using a CRD based approach we would need the following RBAC permissions:
+
+Network plugin - WATCH/LIST/GET:
+
+- Cloud component CRD - as to retrieve the node subnet information from the
+  cloud API
+
+Cloud component - UPDATE/PATCH/CREATE
+
+- Cloud component CRD - as to update its CR with the status of the cloud
+  assignment
+
+**labelled / annotated approach**
+
+Cloud component - UPDATE/PATCH/CREATE
+
+- `node` - as to annotate nodes with information read by the network plugin
+- `events` - as to notify the user of unsuccessful cloud assignment
+
+##### Deployment model
+
+The following section describes a deployment model assuming the simplest
+approach possible: keeping the cloud component as a dedicated OpenShift
+component and using a labelled / annotated approach.
+
+###### OVN-Kubernetes
+
+**Day 0**
+
+1. When OVN-Kubernetes starts: ovnkube-node will read the subnet off of the
+   primary network interface (on AWS and Azure this should correspond to the
+   subnet retrieved from the cloud API) and annotate the node object with
+   `k8s.ovn.org/node-primary-ifaddr` for ovnkube-master to read once egress IP
+   assignments needs to be performed.
+2. The cloud operator issues a `CredentialRequest` to the CCO and reads the
+   secret in the `kube-system` namespace.
+3. The cloud operator will request the nodes subnet from the cloud API.
+4. It will label/annotate the each node object with (for ex:
+   `cloud.api/primary-ifaddr`) containing the subnet information for
+   ovnkube-master to read once egress IP assignment needs to be done. This value
+   would supersede `k8s.ovn.org/node-primary-ifaddr` in the assignment
+   procedure.
+
+**Day X**
```

Comment:
I have a different proposal:

1. User labels nodes for egress assignment
2. User creates EgressIP object
3. ovnkube-master requests that an IP be attached to a node
4. The cloud IP controller attaches the IP, and annotates the node accordingly
5. ovnkube-master waits for the IP to be assigned, and proceeds with enabling it.

Reply:
I would like to avoid interlacing both of these components, to the highest extent possible. Doing this would mean that between 3. 4. and 5. we would have to explicitly code something in ovn-org/ovn-kubernetes that depends on if the feature is running in the cloud or not. In an ideal world I wouldn't change the network plugins behavior in any way except having it look for an annotation on the node object indicating the "node's cloud subnet".